File: InverseAnscombe.py
'''
Created on Oct 7, 2014

@author: haiderriaz
'''
import math
from decimal import *
#from jarray import array , zeros
from array import * 
    

def loadOVSTable():    
    row1 = []
    row2 = []
    with open('Table.dat','r') as f:
        for line in f:
            r1, r2 = line.split(",")
            row1.append(float(r1.strip()))
            row2.append(float(r2.strip())) 
            
    return (row1 , row2)           


# x and y are the keypoints 
# xx are the values you want to use to interpolate
# m = y2 - y1 / x2 - x1
# m*(x2 - x1) = y2 - y1
# y2 = y1 + m*(x2 - x1) <-- x2 is the value you want to use to interpolate, y2 is the output
def interp1(x, y, xx):
  
    out = []
    for pt in xx:
        # Case #1 - if pt < x[0]
        if pt < x[0]:
            
            x1 = x[0]
            x2 = x[1]
            y1 = y[0]
            y2 = y[1]
        # Case #2 - if pt > x[end]
        elif pt > x[-1]:
            
            x1 = x[-2]
            x2 = x[-1]
            y1 = y[-2]
            y2 = y[-1]
        else:
            
            index = binary_search(x, pt) # Figures out where we need to access for the x points
            x1 = x[index]
            x2 = x[index+1]
            y1 = y[index]
            y2 = y[index+1]
    
        # Calculate the slope 
        m = (y2 - y1) / (x2 - x1)
    
        # Interpolate the point
        interp = y1 + m*(pt - x1)
    
        # Add the point to the list
        out.append(interp)
    
    return out


# Inputs - x is an array of values - values are sorted 
# search is the value we want to search for
def binary_search(x, search):
  
    # We are defining our range of searching
    # First is the left side of the range
    first = 0
    # Last is the right side of the range
    last = len(x) - 1
  
    # Calculate middle point
    
  
    # Until we find the number we are looking for
    # or if we don't find what we're looking for
    while (first <= last):
        
        middle = (first + last) / 2
        
    # Access the middle point and see if it's
    # less than what we're looking for 
        if x[middle] < search:
            first = middle + 1 # If the value in the middle is LESS than what we're looking for, move left range up
        elif x[middle] == search: # If the middle value is what we're looking for, get out of the loop
            break
        else: # If the value in the middle is GREATER than what we're looking for, move right range down
            last = middle -1
      
    # If we don't find what we're looking for, return the index of where it SHOULD fit
        if (first > last):
            middle = last
            break
  
        # Returns location of where search is found
      
    return middle







def InvAnscombe(fimg):
    
    (Efz , Ez) = loadOVSTable()

    # Estimation of the asympotic inverse AT (Classical one)
    asympinvfimg = map(lambda x: math.pow((x / 2.0) , 2.0) - (1.0/8.0), fimg) 
    
    
    # Estimation of the optimal inv AT (see paper [2]).
    optimalinvfimg = interp1(Efz , Ez , fimg)
    
    
    # Detection of image areas under and above the validity range of the optimal inverse transform
    maxEfz = max(Efz)
    asympmap = [i for i, e in enumerate(fimg) if e > maxEfz]
    OptInvConstant = 2.0* math.sqrt(3.0/8.0)
    biasedmap = [i for i, e in enumerate(fimg) if e < OptInvConstant]
    
    
    # Filling of areas under and above validity range
    for idx in asympmap:
        optimalinvfimg[idx] = asympinvfimg[idx]
        
    for idx in biasedmap:
        optimalinvfimg[idx] = 0
        
    
    #print optimalinvfimg
    out = array('f' , optimalinvfimg )
    
    
    
    return out    
    
    
    
    

File: InverseAnscombe.java
// Haider Khan - haiderriazkhan@hotmail.com
//
// Optimal inverse Anscombe Transform (AT) 
//
// References
// [1] M. Makitalo and A. Foi, "On the inversion of the Anscombe transformation in low-count Poisson image denoising", 
// Proc. Int. Workshop on Local and Non-Local Approx. in Image Process., LNLA 2009, Tuusula, Finland, pp. 26-32, August 2009.
// [2] M. Makitalo and A. Foi, "Optimal inversion of the Anscombe transformation in low-count Poisson image denoising",
// TIP 2010.
//
// -------------------------------------------------------------------------------





import java.io.BufferedReader;
import java.io.FileReader;
import java.io.IOException;
import ij.IJ;


public class InverseAnscombe {
	
// Load pre-computed correspondence table	
public static Object[] loadOVSTable(){
	
	double[] firstColumn = new double[30000];
	double[] secondColumn = new double[30000];
	int index = 0;

	try {
	      BufferedReader in = new BufferedReader(new FileReader("plugins/CANDLE-J/Table.dat"));
	      String str ;

	      while ((str = in.readLine()) != null) {
	             String[] values = str.split(",");
	             firstColumn[index] = Double.parseDouble(values[0]);
	             secondColumn[index++] = Double.parseDouble(values[1]);           
	      }

	       in.close();
	    } 
	catch (IOException e) 
	{
		IJ.log("Error: Could not load pre-computed correspondence table.");
	}
	
	return new Object[]{firstColumn, secondColumn};
	
}


// Binary Search algorithm for a sorted array of doubles and a floating point search value
public static int binary_search(double[] x, float value){
	
	Integer middle = null;
	int first = 0;
	int last = x.length - 1;
	
	while(first <= last){
		
		middle = (first + last) / 2 ;
		
		if(x[middle] < value){
			
			first = middle + 1;
		}else if(x[middle] == value){
			
			break;
		}else{
			
			last = middle - 1;
		}
		
		if( first > last){
			middle = last;
		    break;
		}
		
	}
	
	return middle;
	
}

// The method implements Matlab's 1D interpolation function
public static float[] interp1(double[] x, double[] y, float[]xq){
	
	double x1,x2,y1,y2,m,interp;
	int index;
	int counter = 0;
	
	float [] vq = new float[xq.length];
	
	for(float pt : xq){
		
		if(Float.isNaN(pt)){
			
			vq[counter++] = Float.NaN;
			continue;
		}
		
		if(pt < x[0]){
			x1 = x[0];
		    x2 = x[1];
		    y1 = y[0];
		    y2 = y[1];
		}else if(pt > x[x.length-1]){
			x1 = x[x.length-2];
		    x2 = x[x.length-1];
		    y1 = y[x.length-2];
		    y2 = y[x.length-1];
			
		}else{
			
			index = binary_search(x, pt);
			
			x1 = x[index];
		    x2 = x[index+1];
		    y1 = y[index];
		    y2 = y[index+1];
			
		}
		
		m = (y2 - y1) / (x2 - x1);
	    interp = y1 + m*(pt - x1);
        
        if(interp < 0){
            vq[counter++] = 0;
            continue;
        }
        
	    vq[counter++] = (float)interp;
		
	}
	
	return vq;
}
	
	
// Parent method that implements the Inverse Anscombe Transform	
public static float[] OVST(float[] fimg){
		
		int len = fimg.length;
		
		float [] asympinvfimg = new float[len];
		
		// Estimation of the asympotic Inverse Anscombe Transform (Classical one)
		for(int i = 0; i < len; i++){
			asympinvfimg[i] =  (float) ( Math.pow( (fimg[i] / 2) , 2 ) - (0.125) );	
		}
		
		Object[] tuple = loadOVSTable();
		
		double[] Efz = (double[]) tuple[0];
		double[] Ez = (double[]) tuple[1];
		
		// Estimation of the optimal Inverse Anscombe Transform (see paper [2]).
		float[] optimalinvfimg = interp1(Efz, Ez , fimg);
		  
		double max_Efz = 100.0025;
		double min_const = 2*Math.sqrt(3/8) ;
		
		// Areas under and above the validity range are filled 
		for(int i=0; i < len; i++){
			
			if( fimg[i] > max_Efz ){
				optimalinvfimg[i] = asympinvfimg[i];
			}else if( fimg[i] < min_const ){
				optimalinvfimg[i] = 0;
			}
			
		}
		
		return optimalinvfimg;
    
		
	}

}


